### ISSUE TITLE
Typeahead documentation adjustments for rxjs 5.5

### ISSUE THREAD

**Issue by andreas-aeschlimann** [ng-bootstrap/ng-bootstrap#2177]
The Typeahead documentation should be adapted to the changes coming in rxjs 5.5.

For example:
```
search = (text$: Observable<string>) =>
    text$
      .debounceTime(200)
      .distinctUntilChanged()
      .map(term => term.length < 2 ? []
        : states.filter(v => v.toLowerCase().startsWith(term.toLocaleLowerCase())).splice(0, 10));
}
```
This does not compile anymore, and could be changed to:
```
search = (text$: Observable<string>) =>
    text$.pipe(
      debounceTime(200),
      distinctUntilChanged(),
      map(term => term.length < 2 ? []
        : states.filter(v => v.toLowerCase().startsWith(term.toLocaleLowerCase())).splice(0, 10)
      )
    );
}
```

**Comment by cristianmartinez** [comment]
This is a good one. Just in case we need more context about the change:
https://github.com/ReactiveX/rxjs/blob/master/doc/pipeable-operators.md

**Comment by king612** [comment]
The new pipe operator pattern is not working for me.  Causes some weird typescript error and will not compile.  I am using:

@angular/cli:  1.6.1
rxjs:  5.5.5
typescript:  2.6.2
@ng-boostrap/ng-bootstrap:  1.0.0

Following the documentation at the link I do:

``import { debounceTime, distinctUntilChanged, map } from 'rxjs/operators';``

and the code in the previous post above.

Visual Studio Code reports no error, but ng build or serve with the --aot switch generates a strange error down in the bowels of typescript:  "TypeError: Cannot read property: 'text' of undefined"


Any help appreciated.  I'm trying to replace ng Material typeahead with ng-bootstrap typeahead.

**Comment by king612** [comment]
Sorry, that is:

``@angular/cli:  1.6.1
@ng-bootstrap/ng-bootstrap:  1.0.0``

**Comment by andreas-aeschlimann** [comment]
This is probably an issue you should talk about on rxjs or angular-cli.

If text is undefined, you probably need to make sure it has an initial value at component creation.

**Comment by king612** [comment]
Well, the issue only occurs with ng-bootstrap.  I have lots of other modules in use in the app, and none of them are experiencing this problem, only ng-bootstrap typeahead.  And the only variable 'text' that is involved is the text$ observable taken directly from your sample code, so someone might want to look into it.

**Comment by king612** [comment]
OK, so my issue was with angular/cli and compiler getting out of sync with the global version.  However, I still can't get a simple typeahead to work with ng-bootstrap.

I'm trying to simply take the text observable and call my service function with it as an argument.  My service function returns the required Observable<any[]>, but the Wikipedia example (which is the closest example to my use case) is frankly way too complex to be useful as a simple example of calling a function that returns an Observable.  Please add a better / simpler example for noobs that does not require sophisticated rxjs expertise.

All I'm trying to do is:

``search = (text$: Observable<string>) => 
    text$.????( term =>
        myService.myMethodThatReturnsObservable<any[]>(term)
    );
``

I've tried map, switchMap, plain old subscribe.  Nothing.  No error message and no typeahead list.  I have verified that my method is returning the appropriate results.

Really shouldn't be this hard.  Frankly, I like the ng Material approach better - just statically bind to an Observable array, and let the developer decide when / how to populate that Observable instead of forcing everything into some complex rxjs chain.  I know it looks impressive, but the goal is to make it easy to use, not to show off our rxjs wizardry.

**Comment by jnizet** [comment]
@pkozlowski-opensource what do you think about using pipeable operators. I think it's a good idea to use them in all the demos, to show the "good way" of using RxJS.

Regarding the code, I see that we're using things like

    _do.call(this._valueChanges, value => {

instead the more traditional

    this._valueChanges.do(value =>

Is it still a good idea to use this style of coding, or should we instead use pipeable operators? As far as I understand, both avoid adding the operator to the Observable prototype, and the pipeable operators are more readable, more idiomatic, and seem to be the recommended way now. See https://www.learnrxjs.io/concepts/operator-imports.html.

If you agree, I can make these changes.

**Comment by maxokorokov** [comment]
I think we should use pipeable operators as it is the recommended way starting from rxjs 5.5. 
However we don't have strictly declared peer dependency on rxjs 5.5 (although angular 5.0.2 kinda has it...)

So I guess we should update demos after declaring it?

Also we should make sure we import them one by one, not from `'rxjs/operators'` (you'll ALL rxjs operators in your bundle with current angular cli / webpack 3):

```typescript
import {tap} from 'rxjs/operators/tap';

this._valueChanges.pipe(tap(value => { /* ... */ }));
``` 

P.S. I played with pipes in `typeahead.ts` and `datepicker.service.ts` and wasn't impressed with changes, so didn't bother to open a PR :)

**Comment by jnizet** [comment]
@maxokorokov are you sure about that? I migrated one of our apps (using the CLI) to pipeable operators, importing them from rxjs/operators,  and didn't notice any significant change in the size of the production bundle. 

It's using RxJS 5.5.7, and Angular CLI 1.7.3. I just built it (with `--prod --sm` flags) to get the source maps, and uses source-map-explorer t inspect the content of the bundle, and got this:

<img width="1253" alt="capture d ecran 2018-03-23 a 21 42 24" src="https://user-images.githubusercontent.com/1608223/37852543-54c6dc34-2ee3-11e8-8e1f-1db70227defd.png">

There are less than 20 operators listed, so its seems all the unused ones are not bundled.

**Comment by maxokorokov** [comment]
@jnizet yep I might have been confused a little. It's looks like it's the case for your own lazy-loaded modules, but shouldn't be a problem for ng-bootstrap.

Ex. using `import { switchMap } from 'rxjs/operators'`, only in one of lazy-loaded modules brings all operators in.

As one includes `NgbXXXModule.forRoot()` in the main app and `NgbXXXModule` in the lazy loaded, I think it should be fine, would still double check it though...

**Comment by maxokorokov** [comment]
Very nice resume, @jnizet !

**Comment by pkozlowski-opensource** [comment]
I believe that all the changes mentioned here have landed in master via #2255, #2256 and subsequent changes. Will be released as part of 2.0.0
